### DISCUSSION TITLE
`numpy` types as instance of `numbers.Real`

### DISCUSSION THREAD

**Discussion by Splines** [microsoft/pyright#7679] [Q&A] [answered]
I've recently asked [this question](https://stackoverflow.com/q/78313586/9655481) on StackOverflow for how to support `numpy` types in a method signature without having to import `numpy` as dependency. It turned out one option is to use the [numeric abstract base class `numbers.Real`](https://docs.python.org/3/library/numbers.html).

```py
>>> import numpy as np
>>> import numbers
>>> isinstance(np.float16(1.2345), numbers.Real)
True
```

However, with the following code:

```py
import numpy as np
import numbers


def cool(my_param: numbers.Real):
    print(my_param + 42)


cool(np.float16(1.2))
```

pyright complains with

```py
error: Argument of type "float16" cannot be assigned to parameter "my_param" of type "Real" in function "cool"
    "float16" is incompatible with "Real" (reportArgumentType)
```

Is there a reason for this behavior?

**Comment by hmc-cs-mdrissi** [comment, marked as answer]
https://stackoverflow.com/questions/69334475/how-to-hint-at-number-types-i-e-subclasses-of-number-not-numbers-themselv https://discuss.python.org/t/numeric-generics-where-do-we-go-from-pep-3141-and-present-day-mypy/17155

are two good resources.

Short answer is numbers related abstract classes do not work at all with type hints and that's case for all type checkers.

**Reply by Splines** [reply]
Oh no, that's unfortunate. Thanks for the very quick reply and the links.

**Reply by erictraut** [reply]
You could define a protocol class that indicates which parts of the interface contract your implementation relies on. As long as numpy satisfies that contract (and is therefore compatible with the protocol), numpy types will be accepted by a type checker.
